src/transport/Dat.ts

```diff
@@ -1,92 +1,254 @@
+                    return resolve(me)
+                }
             })
-            logs.push(dat.key.toString('hex'))
         })
     }
 
-    Download(destinationPath: string, key: any, callback) {
-        this.destination = path.join(__dirname, destinationPath)
-        let dest = this.destination
-        var logs = []
-        var parent = this
-        var envelope = new Envelope()
-        if (!fs.existsSync(dest)) fs.mkdirSync(dest)
-        return DatNode(ram, { key: key, sparse: true }, function (err, dat) {
-            if (err) throw err
-            parent.dat.push(dat)
-            parent.network = dat.joinNetwork(function(err){
-                if (err) throw err
-                if (!dat.network.connected || !dat.network.connecting) {
-                    logs.push('No users currently online for that key.')
-                    callback(logs)
-                    return envelope
+    joinNetwork(callback?:(datnode:DatNode, err?:{}) => void): Promise<DatNode> {
+        let me = this;
+        if (callback) {
+            me.once('join', callback);
+        }
+        return new Promise<DatNode>(function(resolve, reject) {
+            me._dat.joinNetwork(function(err) {
+                if (err) {
+                    me.emit('join', me, err)
+                    console.error('Failed to join network:')
+                    console.error(err)
+                    return reject(err)
+                } else {
+                    me.emit('join', me)
+                    return resolve(me)
+                }
+            });
+        })
+    }
+
+    importFiles(callback?:(datnode:DatNode, err?:{}) => void): Promise<DatNode> {
+        let me = this;
+        if (callback) {
+            me.once('import', callback);
+        }
+        return new Promise<DatNode>(function(resolve, reject) {
+            me._dat.importFiles(function(err) {
+                if (err) {
+                    me.emit('import', me, err)
+                    console.error('Failed to import files:')
+                    console.error(err)
+                    return reject(err)
+                } else {
+                    me.emit('import', me)
+                    return resolve(me)
                 }
             })
-            parent.network.once('connection', function () {
-                logs.push("Connected")
-            })
+        })
+    }
 
-            dat.archive.metadata.update(download)
-
-            function download() {
-                var progress = mirror({ fs: dat.archive, name: '/' }, dest, function (err) {
-                    if (err) throw err
-                    logs.push('Done')
-                    callback(logs)
-                    return envelope
-                })
-                progress.on('put', function (src) {
-                    logs.push('Downloading', src.name)
-                })
+    close(callback?:(datnode:DatNode) => void): Promise<DatNode> {
+        let me = this;
+        if (callback) {
```

Comment:
Question/Request: Since this statement only has one line, are the braces necessary?

```
if (callback) {
  me.once('close', callback);
}
```

to

```
if (callback)
  me.once('close', callback);
```